cpp/include/cugraph/detail/graph_functions.cuh

```diff
@@ -0,0 +1,584 @@
+  rmm::device_uvector<vertex_t> vc_offsets(vertex_count_offsets.size(), handle.get_stream());
+
+  // Transfer data
+  raft::update_device(
+    matrix_partitions.data(), partitions.data(), partitions.size(), handle.get_stream());
+  raft::update_device(major_begin.data(), id_begin.data(), id_begin.size(), handle.get_stream());
+  raft::update_device(minor_end.data(), id_end.data(), id_end.size(), handle.get_stream());
+  raft::update_device(vc_offsets.data(),
+                      vertex_count_offsets.data(),
+                      vertex_count_offsets.size(),
+                      handle.get_stream());
+  raft::update_device(
+    hs_begin.data(), hypersparse_begin.data(), hypersparse_begin.size(), handle.get_stream());
+
+  return std::make_tuple(std::move(matrix_partitions),
+                         std::move(major_begin),
+                         std::move(minor_end),
+                         std::move(hs_begin),
+                         std::move(vc_offsets));
+}
+/**
+ * @brief Gather valid edges present on the current gpu
+ *
+ * Collect all the edges that are present in the adjacency lists on the current gpu
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @tparam EdgeIndexIterator Type of the iterator for edge indices.
+ * @tparam GPUIdIterator  Type of the iterator for gpu id identifiers.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param active_majors_in_row Device vector containing all the vertex id that are processed by
+ * gpus in the column communicator
+ * @param active_major_gpu_ids Device vector containing the gpu id associated by every vertex
+ * present in active_majors_in_row
+ * @param edge_index_first Iterator pointing to the first destination index
+ * @param indices_per_source Number of indices supplied for every source in the range
+ * [vertex_input_first, vertex_input_last)
+ * @param global_degree_offset Global degree offset to local adjacency list for every source
+ * represented by current gpu
+ * @return A tuple of device vector containing the majors, minors and gpu_ids gathered locally
+ */
+template <typename GraphViewType, typename EdgeIndexIterator, typename gpu_t>
+std::tuple<rmm::device_uvector<typename GraphViewType::vertex_type>,
+           rmm::device_uvector<typename GraphViewType::vertex_type>,
+           rmm::device_uvector<gpu_t>>
+gather_local_edges(
+  raft::handle_t const& handle,
+  GraphViewType const& graph_view,
+  const rmm::device_uvector<typename GraphViewType::vertex_type>& active_majors_in_row,
+  const rmm::device_uvector<gpu_t>& active_major_gpu_ids,
+  EdgeIndexIterator edge_index_first,
+  size_t indices_per_major,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_degree_offsets)
```

Comment:
Should we expose all these implementation details in the header file under `cpp/include`?

Cant' we just expose the header for `gather_local_edges` and place implementations under `src/sampling/gather_local_edges_impl.cuh` (or better place/naming under the `src` directory)? I assume users will just call `gather_local_edges`???